File: incre-tests/synduce/misc/simple_nnf.f
Inductive Formula = flit Bool | fand {Formula, Formula} | forr {Formula, Formula} | fnot Formula;
Inductive NnfFormula = nfneglit Bool | nflit Bool | nfand {NnfFormula, NnfFormula} | nfor {NnfFormula, NnfFormula};

spec = fix (
  \f: Formula -> Bool. \x: Formula.
  match x with
    flit b -> b
  | fand {b1, b2} -> and (f b1) (f b2)
  | forr {b1, b2} -> or (f b1) (f b2)
  | fnot b -> if (f b) then false else true
  end
);

repr = fix (
  \f: NnfFormula -> Compress Formula. \x: NnfFormula.
  match x with
    nflit b -> flit b
  | nfneglit b -> fnot (flit b)
  | nfand {b1, b2} -> fand {f b1, f b2}
  | nfor {b1, b2} -> forr {f b1, f b2}
  end
);

main = \xs: NnfFormula. spec (repr xs);

File: ml-benchmark/synduce/misc/simple_nnf.ml
type formula =
  | FLit of bool
  | FAnd of formula * formula
  | FOr of formula * formula
  | FNot of formula

let rec run = function
  | FLit b -> b
  | FAnd (x, y) -> run x && run y
  | FOr (x, y) -> run x || run y
  | FNot x -> not (run x)

type nnf_formula =
  | NFNegLit of bool
  | NFLit of bool
  | NFAnd of nnf_formula * nnf_formula
  | NFOr of nnf_formula * nnf_formula

val repr: nnf_formula -> formula compress 
let rec repr = function
  | NFLit b -> FLit b
  | NFNegLit b -> FNot (FLit b)
  | NFAnd (x, y) -> FAnd (repr x, repr y)
  | NFOr (x, y) -> FOr (repr x, repr y)

let prog e = run (repr e)